File: java/src/main/java/com/mlt/converter/geometry/SpaceFillingCurve.java
package com.mlt.converter.geometry;

public abstract class SpaceFillingCurve {
  protected int tileExtent;
  protected int numBits;
  protected int coordinateShift;
  private final int minBound;
  private final int maxBound;

  public SpaceFillingCurve(int minVertexValue, int maxVertexValue) {
    // TODO: fix tile buffer problem
    /* Each tile can have a buffer around, which means the coordinate values are extended beyond the specified tileExtent.
     * Currently we are extending size tile size be half of to the size into each direction, which works well for the test tilesets.
     * But this leads to problems if the tile coordinates are not within this boundaries.
     * */
    coordinateShift = minVertexValue < 0 ? Math.abs(minVertexValue) : 0;
    this.tileExtent = maxVertexValue + coordinateShift;
    this.numBits = (int) Math.ceil((Math.log(this.tileExtent + 1) / Math.log(2)));
    ;
    minBound = minVertexValue;
    maxBound = maxVertexValue;
  }

  protected void validateCoordinates(Vertex vertex) {
    // TODO: also check of int overflow as we limiting the sfc ids to max int size
    if (vertex.x() < minBound
        || vertex.y() < minBound
        || vertex.x() > maxBound
        || vertex.y() > maxBound) {
      // System.err.println("The specified tile buffer size is currently not supported.");
      throw new IllegalArgumentException(
          "The specified tile buffer size is currently not supported.");
    }
  }

  public abstract int encode(Vertex vertex);

  public abstract int[] decode(int mortonCode);

  public int numBits() {
    return this.numBits;
  }

  public int coordinateShift() {
    return this.coordinateShift;
  }
}

File: ts/src/vector/geometry/spaceFillingCurve.ts

export default abstract class SpaceFillingCurve {
    protected tileExtent: number;
    protected _numBits: number;
    protected _coordinateShift: number;
    private readonly minBound: number;
    private readonly maxBound: number;

    constructor(minVertexValue: number, maxVertexValue: number) {
        // TODO: fix tile buffer problem
        this._coordinateShift = minVertexValue < 0 ? Math.abs(minVertexValue) : 0;
        this.tileExtent = maxVertexValue + this._coordinateShift;
        this._numBits = Math.ceil(Math.log2(this.tileExtent));
        this.minBound = minVertexValue;
        this.maxBound = maxVertexValue;
    }

    protected validateCoordinates(vertex: { x: number, y: number }): void {
        // TODO: also check for int overflow as we are limiting the sfc ids to max int size
        if (vertex.x < this.minBound || vertex.y < this.minBound || vertex.x > this.maxBound || vertex.y > this.maxBound) {
            throw new Error("The specified tile buffer size is currently not supported.");
        }
    }

    abstract encode(vertex: { x: number, y: number }): number;

    abstract decode(mortonCode: number): { x: number, y: number };

    numBits(): number {
        return this._numBits;
    }

    coordinateShift(): number {
        return this._coordinateShift;
    }
}
